src/anyio/functools.py

```diff
@@ -0,0 +1,253 @@
+from __future__ import annotations
+
+__all__ = ("cache", "lru_cache", "reduce")
+
+import functools
+import sys
+from collections import OrderedDict, defaultdict
+from collections.abc import (
+    AsyncIterable,
+    Awaitable,
+    Callable,
+    Coroutine,
+    Hashable,
+    Iterable,
+)
+from functools import update_wrapper
+from inspect import iscoroutinefunction
+from typing import Any, Generic, NamedTuple, TypedDict, TypeVar, cast, overload
+
+from ._core._synchronization import Lock
+from .lowlevel import RunVar
+
+if sys.version_info >= (3, 11):
+    from typing import ParamSpec
+else:
+    from typing_extensions import ParamSpec
+
+T = TypeVar("T")
+S = TypeVar("S")
+P = ParamSpec("P")
+lru_cache_items: RunVar[
+    defaultdict[Callable[..., Any], tuple[OrderedDict[Hashable, Any], Lock]]
+] = RunVar("lru_cache_items")
+
+
+class _InitialMissingType:
+    pass
+
+
+initial_missing: _InitialMissingType = _InitialMissingType()
+
+
+class AsyncCacheInfo(NamedTuple):
+    hits: int
+    misses: int
+    maxsize: int | None
+    currsize: int
+
+
+class AsyncCacheParameters(TypedDict):
+    maxsize: int | None
+    typed: bool
+
+
+class AsyncLRUCacheWrapper(Generic[P, T]):
+    def __init__(
+        self, func: Callable[..., Awaitable[T]], maxsize: int | None, typed: bool
+    ):
+        self.__wrapped__ = func
+        self._hits: int = 0
+        self._misses: int = 0
+        self._maxsize = max(maxsize, 0) if maxsize is not None else None
+        self._currsize: int = 0
+        self._typed = typed
+        update_wrapper(self, func)
+
+    def cache_info(self) -> AsyncCacheInfo:
+        return AsyncCacheInfo(self._hits, self._misses, self._maxsize, self._currsize)
+
+    def cache_parameters(self) -> AsyncCacheParameters:
+        return AsyncCacheParameters(maxsize=self._maxsize, typed=self._typed)
+
+    def cache_clear(self) -> None:
```

Comment:
Perhaps add a `__del__` that calls `cache_clear`?

Reply:
OK, also done now.